File: 239-sliding-window-maximum/sliding-window-maximum.java
import java.util.*;

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {

        List<Integer> res = new ArrayList<>();
        Deque<Integer> q = new ArrayDeque<>(); // stores indices

        int l = 0;

        for (int r = 0; r < nums.length; r++) {

            // remove smaller elements from back
            while (!q.isEmpty() && nums[r] > nums[q.peekLast()]) {
                q.pollLast();
            }

            q.offerLast(r);

            // remove left element if out of window
            if (l > q.peekFirst()) {
                q.pollFirst();
            }

            // window formed
            if (r + 1 >= k) {
                res.add(nums[q.peekFirst()]);
                l++;
            }
        }

        // convert List to array
        int[] ans = new int[res.size()];
        for (int i = 0; i < res.size(); i++) {
            ans[i] = res.get(i);
        }

        return ans;
    }
}

File: 239-sliding-window-maximum/sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        ans = []

        heap =[]

        for r in range(len(nums)):
            heapq.heappush(heap , (-nums[r] , r))

            if r >= k - 1:

                while heap[0][1] <= r - k:
                    heapq.heappop(heap)

                ans.append(-heap[0][0])
        return ans






        # res = []

        # q = deque() #INDEX 
        # l = r = 0

        # while r < len(nums):

        #     while q and nums[r] > nums[q[-1]]:
        #         q.pop()

        #     q.append(r)

        #     if l > q[0]:
        #         q.popleft()
            
        #     if (r + 1) >= k:
        #         res.append(nums[q[0]])
        #         l += 1
        #     r += 1
        # return res
        

            

